lib/abstract_format.ml

```diff
@@ -356,6 +369,40 @@ and expr_of_sf sf =
   | Sf.Tuple (4, [Sf.Atom "case"; Sf.Integer line; e; Sf.List clauses]) ->
      ExprCase (line, e |> expr_of_sf, clauses |> List.map cls_of_sf)
 
+  (* a local function reference *)
+  | Sf.Tuple (3, [Sf.Atom "fun";
+                  Sf.Integer line;
+                  Sf.Tuple (3, [Sf.Atom "function";
+                                Sf.Atom name;
+                                Sf.Integer arity])]) ->
+    ExprLocalFunRef (line, name, arity)
+
+  (* a remote function reference *)
+  | Sf.Tuple (3, [Sf.Atom "fun";
+                  Sf.Integer line;
+                  Sf.Tuple (4, [Sf.Atom "function";
+                                module_name;
+                                function_name;
+                                arity])]) ->
+    ExprRemoteFunRef (line,
+                      module_name |> module_name_atom_or_variable_of_sf,
```

Comment:
I prefer to use `expr_of_sf` instead of `module_name_atom_or_variable_of_sf` and check a kind of terms by pattern matching for `expr_t`.

Reasons:
- There are no groups such as `module_name_atom_or_variable` in the [official document](http://erlang.org/doc/apps/erts/absform.html#expressions)
- Reduce redundant codes
  - `ModuleNameAtom` can be replaced by `ExprLit` and `LitAtom`
  - `ModuleNameVariable` can be also replaced by `ExprVar`

I have a same opinion also for `function_name_atom_or_variable_of_sf` and `arity_integer_or_variable_of_sf`.

What do you think about this?

Reply:
I think that constraints for terms of absform which you wrote are correct! :+1: https://github.com/erlang/otp/blob/master/lib/stdlib/src/erl_parse.yrl#L468

Reply:
I think `expr_t` is too general in this case.
I mean, using `expr_t` forces obeam users to handle unnecessary case like following:

```
match module_name with
| Abstract_format.ExprLit (Abstract_format.LitAtom (_, name)) -> ...
| Abstract_format.ExprVar (_, var) -> ...
| _ -> raise InvalidModuleName (* unnecessary case! this clause is needed when we use expr_t *)
```

This is why I introduced `ModuleNameAtom`, `ModuleNameVariable`, and so on.

As you can see, `ModuleNameAtom` is an awful and redundant name.
We should use more precise names for these constructors.

Reply:
FYI: the constraints of the syntax of the function expression is documented in the [official reference manual](http://erlang.org/doc/reference_manual/expressions.html#fun-expressions)

Reply:
I see... However, The [document about absform](http://erlang.org/doc/apps/erts/absform.html#expressions) does not restrict a kind of terms :crying_cat_face: (So I think Erlang reserves room for extensions of the syntax at least on absform.)

> FYI: the constraints of the syntax of the function expression is documented in the official reference manual

Thanks!

Any other opinions? @amutake

Reply:
Hmm... I agree with @youxkei 's opinion.
I think such pattern restriction makes obeam users happy even if Erlang reserves future extensions.

Reply:
ok. Then, please rename them to more generic name which suitable for the absform doc :pray:

Reply:
I renamed the constructors according to the https://github.com/erlang/otp/blob/71215060abff81b61844f8517bde72a2252cef53/lib/stdlib/src/erl_parse.yrl#L468.